pkg/controller/istiocsr/networkpolicies.go

```diff
@@ -0,0 +1,108 @@
+		if err != nil {
+			return fmt.Errorf("failed to get network policy from asset %s: %w", assetPath, err)
+		}
+		if err := r.createOrUpdateNetworkPolicy(obj, istioCSRCreateRecon); err != nil {
+			return fmt.Errorf("failed to create/update network policy from %s: %w", assetPath, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) getNetworkPolicyFromAsset(assetPath string, istiocsr *v1alpha1.IstioCSR, resourceLabels map[string]string) (*networkingv1.NetworkPolicy, error) {
+	// Get the target namespace for istio-csr deployment
+	namespace := istiocsr.GetNamespace()
+	if namespace == "" {
+		namespace = istiocsr.Spec.IstioCSRConfig.Istio.Namespace
+	}
+
+	// Read the asset and decode it
+	assetBytes := assets.MustAsset(assetPath)
+	obj, err := runtime.Decode(codecs.UniversalDeserializer(), assetBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode network policy asset %s: %w", assetPath, err)
+	}
+
+	policy := obj.(*networkingv1.NetworkPolicy)
+
+	// Set the correct namespace
+	policy.Namespace = namespace
+
+	// Merge resource labels
+	if policy.Labels == nil {
+		policy.Labels = make(map[string]string)
+	}
+	for k, v := range resourceLabels {
+		policy.Labels[k] = v
+	}
+
+	return policy, nil
+}
+
+func (r *Reconciler) createOrUpdateNetworkPolicy(policy *networkingv1.NetworkPolicy, istioCSRCreateRecon bool) error {
+	desired := policy.DeepCopy()
+	policyName := fmt.Sprintf("%s/%s", desired.GetNamespace(), desired.GetName())
+	r.log.V(4).Info("reconciling network policy resource", "name", policyName)
+
+	fetched := &networkingv1.NetworkPolicy{}
+	key := types.NamespacedName{
+		Name:      desired.GetName(),
+		Namespace: desired.GetNamespace(),
+	}
+	exist, err := r.Exists(r.ctx, key, fetched)
+	if err != nil {
+		return FromClientError(err, "failed to check %s network policy resource already exists", policyName)
+	}
+
+	if exist && istioCSRCreateRecon {
+		r.eventRecorder.Eventf(nil, corev1.EventTypeWarning, "ResourceAlreadyExists", "%s network policy resource already exists, maybe from previous installation", policyName)
+	}
+	if exist && hasObjectChanged(desired, fetched) {
+		r.log.V(1).Info("network policy has been modified, updating to desired state", "name", policyName)
+		if err := r.UpdateWithRetry(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to update %s network policy resource", policyName)
+		}
+		r.eventRecorder.Eventf(nil, corev1.EventTypeNormal, "Reconciled", "network policy resource %s reconciled back to desired state", policyName)
+	} else {
+		r.log.V(4).Info("network policy resource already exists and is in expected state", "name", policyName)
+	}
+	if !exist {
+		if err := r.Create(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to create %s network policy resource", policyName)
+		}
+		r.eventRecorder.Eventf(nil, corev1.EventTypeNormal, "Reconciled", "network policy resource %s created", policyName)
```

Comment:
Events shouldn't have nil for object.
```suggestion
		r.eventRecorder.Eventf(istiocsr, corev1.EventTypeNormal, "Reconciled", "network policy resource %s created", policyName)
```